iot/end_to_end_example/device/app/src/main/kotlin/com/example/androidthings/fancontrol/FanControlActivity.kt

```diff
@@ -0,0 +1,379 @@
+                        .setDeviceId(BuildConfig.DEVICE_ID)
+                        .build()
+
+                // Initialize the IoT Core client
+                client = IotCoreClient.Builder()
+                        .setConnectionParams(connectionParams)
+                        .setKeyPair(keys)
+                        .setOnConfigurationListener(OnConfigurationListener { this.onConfigurationReceived(it) })
+                        .build()
+
+                // Connect to Cloud IoT Core
+                client!!.connect()
+
+                mHandler.post(mTempReportRunnable)
+            }
+        } catch (ikse: InvalidKeySpecException) {
+            Log.e(TAG, "INVALID Key spec", ikse)
+        } catch (nsae: NoSuchAlgorithmException) {
+            Log.e(TAG, "Algorithm not supported", nsae)
+        } catch (ioe: IOException) {
+            Log.e(TAG, "Could not load key from file", ioe)
+        }
+    }
+
+    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+        if (keyCode == KeyEvent.KEYCODE_A) {
+            Log.d(TAG, "A Click")
+            return true
+        }
+        return super.onKeyUp(keyCode, event)
+    }
+
+    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
+        return if (keyCode == KeyEvent.KEYCODE_A) {
+            true
+        } else super.onKeyUp(keyCode, event)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
+        if (mButtonInputDriver != null) {
+            try {
+                mButtonInputDriver!!.close()
+            } catch (e: IOException) {
+                e.printStackTrace()
+            }
+
+            mButtonInputDriver = null
+        }
+
+        if (mLedstrip != null) {
+            try {
+                mLedstrip!!.brightness = 0
+                mLedstrip!!.write(IntArray(7))
+                mLedstrip!!.close()
+            } catch (e: IOException) {
+                Log.e(TAG, "Error disabling ledstrip", e)
+            } finally {
+                mLedstrip = null
+            }
+        }
+
+        if (mLed != null) {
+            try {
+                mLed!!.value = false
+                mLed!!.close()
+            } catch (e: IOException) {
+                Log.e(TAG, "Error disabling led", e)
+            } finally {
+                mLed = null
+            }
+        }
+
+        // clean up Cloud publisher.
+        if (client != null && client!!.isConnected) {
+            client!!.disconnect()
+        }
+    }
+
+    companion object {
+
+        private val TAG = FanControlActivity::class.java.getSimpleName()
+        private val LEDSTRIP_BRIGHTNESS = 1
+
+        private val MSG_UPDATE_BAROMETER_UI = 1
+
+        @Throws(IOException::class)
+        private fun inputStreamToBytes(inputStream: InputStream): ByteArray {
+            val buffer = ByteArrayOutputStream()
+            var nRead: Int
+            val data = ByteArray(16384)
+
+            do {
```

Comment:
i thought there are better methods for kotlin to copy stream... `copyTo`?

Reply:
also, if you are looking for byte array, `readBytes`?

Reply:
yeesh, good catch :D.

fixed!